### PR TITLE
[Kimi K3] Quantile Balancing for the MoE router bias

### PR DESCRIPTION
**Pull request by QIU023**
## Summary

Add quantile balancing for Kimi K3's MoE router bias.

- The bias is solved at each optimizer step from an accumulated per-expert load histogram at the balanced quantile (report sec. 2.3.3, eqs. 13-14, app. D), instead of nudged by a fixed sign step from the current step's counts. Same bias tensor, same hook point, no model change.
- The hook is one preallocated `(num_experts, num_bins)` histogram per MoE layer and a single `add_` on the router's scores; the solve inverts the accumulated CDF at the Top-(k+1) cutoff with mean-centred margins.
- The `kimi_k3_debugmodel_qb` flavor installs the solver as the model spec's `post_optimizer_build_fn`, the slot core's sign-rule hook occupies; under the flavor the solver is the only writer of `expert_bias_E`.

## Implementation

- Histograms use a fixed score range `[-1, 1]` with 512 bins and mirrored margins; the report's app. D uses an adaptive range and 1000 bins. On one frozen batch the exact-quantile solve and the histogram solve agree within 0.02 in load cv.
- Histograms are summed once over the loss mesh (dp_replicate x dp_shard x cp), stacked across layers into one collective, and not over tp: the router's scores are Replicate under TP, and a second sum would scale the histogram by the tp degree.
- Branch-free and allocation-free: selective activation checkpointing replays the router forward, and a lazily allocated buffer changes the op sequence between forward and recompute (observed; with this shape SAC on and off print identical losses). Full-AC replays count every token twice, and `cumsum / total` is invariant to a uniform factor.
- The solver resets the MoE's per-step token counter, which core's hook did.

## Limitations

- The flavor replaces core's load-balancing registration instead of composing with it: a model spec holds one `post_optimizer_build_fn`.
- The histogram range is fixed: scores outside `[-1, 1]` land in the end bins. The debug model's scores stay inside it.
- Not exercised with pipeline parallelism.

## Tests

```text
pytest tests/unit_tests/cpu/test_quantile_balance.py
```

Result: `14 passed` (histogram accumulation, CDF inversion, the solve against a brute-force reference, in-place zeroing, the SAC-identical op sequence). `tests/unit_tests/gpu/test_kimi_k3.py`: `2 passed, 1 skipped`.

The deterministic BF16 comparison used `seed=42`, `--debug.deterministic`, one seed checkpoint, 8192 tokens per rank per step in 256-token micro-batches, and 10 training steps. The sign-step reference is the debug flavor on the parent commit `65ba8a697`; quantile balancing is the qb flavor on this commit. Percentages are relative to the sign-step run of the same parallelism; dp8 x ep8 is the K3 layout (32 experts, 4 per rank, the histogram summed over the 8-rank loss mesh).

Step 1 is identical under both hooks by construction (the bias is 0 before the first solve); from step 2 the two hooks route the same tokens to different experts, so every row below step 1 is two different runs of the same model rather than a numerics comparison, and the percentages size that divergence rather than an error.

| config | hook | step 1 | step 3 | step 10 |
| --- | --- | ---: | ---: | ---: |
| dp8 x ep8 | sign-step (main) | `12.522390` | `6.618530` | `2.754520` |
| dp8 x ep8 | quantile balancing | `12.522390` (`0%`) | `6.791680` (`2.62%`) | `2.743960` (`0.38%`) |
| dp2 x ep2 | sign-step (main) | `12.523720` | `7.502180` | `3.211300` |
| dp2 x ep2 | quantile balancing | `12.523720` (`0%`) | `7.284180` (`2.91%`) | `3.156020` (`1.72%`) |
| dp1 | sign-step (main) | `12.518870` | `7.112520` | `3.113010` |
| dp1 | quantile balancing | `12.518870` (`0%`) | `7.353260` (`3.38%`) | `3.047750` (`2.10%`) |

Gradient norms on the two configurations that carry them:

| config | hook | step 1 | step 3 | step 10 |
| --- | --- | ---: | ---: | ---: |
| dp8 x ep8 | sign-step (main) | `13.3125` | `8.6875` | `1.1953` |
| dp8 x ep8 | quantile balancing | `13.3125` (`0%`) | `10.875` (`25.2%`) | `1.3125` (`9.81%`) |
| dp1 | sign-step (main) | `14.125` | `10.0625` | `2.0312` |
| dp1 | quantile balancing | `14.125` (`0%`) | `9.5` (`5.59%`) | `2.1406` (`5.39%`) |

As a control, the sign-step flavor on this commit matched the parent commit's run at every step (loss and grad norm), with and without the load probe that produced the table below. Every rank holds the same solved bias after every step (all-gathered and compared, ep8 included).

What the change is for is the load. Expert load per MoE layer (tokens routed to each expert, summed over the loss mesh) as the coefficient of variation over the 32 experts and the largest and smallest expert's load relative to the mean, averaged over the 23 MoE layers and over the step window; the bias range at step 10:

| config | hook | cv steps 1-5 | cv 6-10 | max/mean 6-10 | min/mean 6-10 | bias range at 10 |
| --- | --- | ---: | ---: | ---: | ---: | --- |
| dp8 x ep8 | sign-step (main) | 1.25 | 1.49 | 5.5 | 0.01 | [-0.016, 0.006] |
| dp8 x ep8 | quantile balancing | 1.09 | 0.82 | 3.4 | 0.05 | [-0.297, 0.333] |
| dp2 x ep2 | sign-step (main) | 1.26 | 1.51 | 5.7 | 0.01 | [-0.016, 0.006] |
| dp2 x ep2 | quantile balancing | 1.13 | 0.90 | 3.6 | 0.03 | [-0.352, 0.339] |
| dp2 | sign-step (main) | 1.25 | 1.54 | 5.9 | 0.02 | [-0.016, 0.006] |
| dp2 | quantile balancing | 1.10 | 0.93 | 3.7 | 0.03 | [-0.400, 0.341] |
| dp1 | sign-step (main) | 1.20 | 1.36 | 5.1 | 0.02 | [-0.016, 0.006] |
| dp1 | quantile balancing | 1.01 | 0.80 | 3.2 | 0.05 | [-0.291, 0.344] |

With the router frozen (`lr` 0, dp1, 10 steps, the same data through unchanging weights) the sign step moves the load cv from 1.05 (steps 1-5) to 0.89 (steps 6-10) and quantile balancing from 0.76 to 0.58: the hook moves the loads, not the training.

Runs on RTX 5060 Ti (SM120) with the KDA capability check widened locally so Attention Gym's Triton path runs; the flavor is otherwise unchanged.

### CODE REVIEW

**Comment by QIU023** [comment]
Hi Shuhua @shuhuayu , as we talked in the tracking list, the QB PR is formally raised for your review as a part of P0 features, thank you very much!

File: torchtitan/models/kimi_k3/config_registry.py [hunk 1]
```diff
@@ -61,6 +62,8 @@ def _kimi_k3_multimodal_dataloader(
 
 def kimi_k3_debugmodel() -> Trainer.Config:
     model_spec = model_registry("debugmodel")
+    # The report balances the router bias by solving it, not by stepping it.
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/config_registry.py:65]: nit: remove.

File: torchtitan/models/kimi_k3/config_registry.py [hunk 2]
```diff
@@ -92,3 +95,4 @@ def kimi_k3_debugmodel() -> Trainer.Config:
         ),
         activation_checkpoint=SelectiveAC.Config(),
     )
+
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/config_registry.py:98]: nit: remove.

File: torchtitan/models/kimi_k3/config_registry.py [hunk 3]
```diff
@@ -61,6 +62,8 @@ def _kimi_k3_multimodal_dataloader(
 
 def kimi_k3_debugmodel() -> Trainer.Config:
     model_spec = model_registry("debugmodel")
+    # The report balances the router bias by solving it, not by stepping it.
+    model_spec.post_optimizer_build_fn = register_quantile_balancing
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/config_registry.py:66]: should this go to `model_registry`? should apply to all k3 models, not only the debug model variant.

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        self._handles.clear()
+
+
+def register_quantile_balancing(
+    optimizers, model_parts, parallel_dims, *, num_bins: int = 512
+) -> QuantileBalancer:
+    """``post_optimizer_build_fn`` that replaces the sign rule with QB.
+
+    The slot holds one function, so registering this one leaves core's
+    ``register_moe_load_balancing_hook`` unregistered.
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/quantile_balance.py:290]: change this after moving `post_optimizer_build_fn=register_quantile_balancing` to model registry.

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    """
+    loss_mesh = parallel_dims.get_optional_mesh("loss")
+    balancer = QuantileBalancer(
+        model_parts,
+        num_bins=num_bins,
+        loss_group=None if loss_mesh is None else loss_mesh.get_group(),
+    )
+    optimizers.register_step_pre_hook(lambda *a, **kw: balancer.step())
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/quantile_balance.py:298]:
```suggestion
    optimizers.register_step_pre_hook(lambda *args, **kwargs: balancer.step())
```

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+    size (896 experts, 512 bins, 92 layers), traded against resolution.
+    """
+
+    def __init__(
+        self,
+        model_parts,
+        *,
+        num_bins: int = 512,
+        lo: float = -1.0,
+        hi: float = 1.0,
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/quantile_balance.py:164]: it is inconsistent with the report to keep them the same layer by layer, these values are derived from current step bias.

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    per-step solve.
+    """
+    ids, _ = topk_with_cutoff(scores_TE, bias_E, top_k)
+    return torch.bincount(ids.reshape(-1), minlength=scores_TE.size(-1))
+
+
+# ----- Runtime integration ------------------------------------------------ #
+# A forward hook on each router recomputes Top-(k+1) for the cutoff and adds to
+# the layer's histogram; the step pools them over the loss mesh and overwrites
+# expert_bias_E, which makes this the only writer of the bias.
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/quantile_balance.py:148]: nit: remove.

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+
+class QuantileBalancer:
+    """Drives Quantile Balancing over a training run.
+
+    One ``(E, num_bins)`` int32 histogram per MoE layer: 169 MiB at K3's full
+    size (896 experts, 512 bins, 92 layers), traded against resolution.
+    """
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/quantile_balance.py:156]: nit: remove.

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+    def _iter_moes(model_parts):
+        from torchtitan.models.common.moe import MoE
+
+        for part in model_parts:
+            for m in part.modules():
+                if isinstance(m, MoE):
+                    yield m
+
+    def _make_hook(self, key: int):
+        def hook(router, args, output):
```
Comment (JavaZeroo) [inline torchtitan/models/kimi_k3/quantile_balance.py:209]: This hook executed on every router forward, including eval. I think this should only be executed when `self.training == True`. Otherwise it will effect the next training step when eval.

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+            keys = list(self._counts)
+            stacked = torch.stack([self._counts[k] for k in keys])
+            dist.all_reduce(stacked, group=self.loss_group, op=dist.ReduceOp.SUM)
+            for i, key in enumerate(keys):
+                self._counts[key].copy_(stacked[i])
+
+        for key, counts in self._counts.items():
+            bias = quantile_balance_bias_histogram(
+                counts.to(torch.int64), self._top_k[key], lo=self.lo, hi=self.hi
+            )
```
Comment (JavaZeroo) [inline torchtitan/models/kimi_k3/quantile_balance.py]: Seems self.lo / self.hi are fixed at -1 / 1 here, but App. D of the K3 report recomputes the range every step as [b_min - 1, b_max + 1] from the current bias. The margins live in (-1 - max(b), 1 - min(b)), so once the bias spreads past the fixed range the clamped values pile into the end bins and the quantile read from there is wrong.

File: torchtitan/models/kimi_k3/quantile_balance.py (hunk 9 above)
Comment (QIU023) [reply torchtitan/models/kimi_k3/quantile_balance.py:209]: Thanks! will address tomorrow

File: torchtitan/models/kimi_k3/quantile_balance.py [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+            bias_E = self._moes[key].expert_bias_E
+            assert isinstance(bias_E, torch.Tensor)
+            with torch.no_grad():
+                # Under TP the gate's output is a DTensor and expert_bias_E is
+                # not; the scores are Replicate there, so to_local is exact.
+                if isinstance(scores_BLE, DTensor):
+                    scores_BLE = scores_BLE.to_local()
+                scores_TE = scores_BLE.detach().reshape(-1, scores_BLE.size(-1))
+                bias = bias_E.to_local() if isinstance(bias_E, DTensor) else bias_E
+                _, cutoff_T = topk_with_cutoff(
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/quantile_balance.py:221]: top_k is done twice one in forward one in hook.

**Review by shuhuayu — COMMENTED** [review]
Found some gaps in this pr that are inconsistent with the report and some abstractions can be enhanced. Since this is p0, decided to move with a pr from titan side. https://github.com/pytorch/torchtitan/pull/4577. You may help review.

**Comment by QIU023** [comment]
NP, since Shuhua has taken over in another PR, closing this and moving to review in the new PR
